pandas/tools/merge.py

```diff
@@ -1367,7 +1367,10 @@ def __init__(self, objs, axis=0, join='outer', join_axes=None,
                 clean_keys.append(k)
                 clean_objs.append(v)
             objs = clean_objs
-            keys = clean_keys
+            if isinstance(keys, cat.CategoricalIndex):
```

Comment:
needs a rebase, but I think this would almost work as-is, as IIRC we do something like
``keys = _ensure_index(keys)`` already